python/cudf/cudf/core/abc.py

```diff
@@ -17,6 +21,36 @@
     import pickle
 
 
+_tls = threading.local()
+
+
+def _get_tls_stream():
+    stream = getattr(_tls, "stream", None)
+    if stream is None:
+        _tls.stream = stream = cupy.cuda.Stream(non_blocking=True)
+    return stream
+
+
+def _ensure_on_host(is_cuda, buf):
+    if is_cuda:
+        with _get_tls_stream() as stream:
+            return memoryview(cupy.asarray(buf).get(stream=stream))
+    else:
+        return memoryview(buf)
+
+
+def _ensure_maybe_on_device(is_cuda, buf):
+    buf = memoryview(buf)
+    if is_cuda:
+        with _get_tls_stream() as stream:
+            #dbuf = rmm.DeviceBuffer(size=buf.nbytes, stream=stream.ptr)
+            dbuf = rmm.DeviceBuffer(size=buf.nbytes)
```

Comment:
Ideally we would use the `stream.ptr` for allocation of the `DeviceBuffer` here. However when doing that we got a `cudaErrorInvalidResourceHandle` exception (shown in detail below).

Nothing stuck out to me when looking at [the CuPy `Stream` creation code]( https://github.com/cupy/cupy/blob/7f401f0cb67a36ce87ddb5aeb18d38959c9c5b06/cupy/cuda/stream.pyx#L273-L279 ). That said, maybe this isn't too important if this code was rewritten in Cython in RMM or a PTDS-enabled RMM build were used.

In any event, the key takeaway is that deserialization may not be as fast as we would like due to the use of the default stream here.

<details>

```python
/datasets/jkirkham/miniconda/envs/rapids15dev/lib/python3.8/site-packages/cudf/core/abc.py in host_deserialize(cls, header, frames)
    156         """
    157         with ThreadPoolExecutor() as executor:
--> 158             frames = list(
    159                 executor.map(
    160                     _ensure_maybe_on_device, header["is-cuda"], frames

/datasets/jkirkham/miniconda/envs/rapids15dev/lib/python3.8/concurrent/futures/_base.py in result_iterator()
    609                     # Careful not to keep a reference to the popped future
    610                     if timeout is None:
--> 611                         yield fs.pop().result()
    612                     else:
    613                         yield fs.pop().result(end_time - time.monotonic())

/datasets/jkirkham/miniconda/envs/rapids15dev/lib/python3.8/concurrent/futures/_base.py in result(self, timeout)
    430                 raise CancelledError()
    431             elif self._state == FINISHED:
--> 432                 return self.__get_result()
    433 
    434             self._condition.wait(timeout)

/datasets/jkirkham/miniconda/envs/rapids15dev/lib/python3.8/concurrent/futures/_base.py in __get_result(self)
    386     def __get_result(self):
    387         if self._exception:
--> 388             raise self._exception
    389         else:
    390             return self._result

/datasets/jkirkham/miniconda/envs/rapids15dev/lib/python3.8/concurrent/futures/thread.py in run(self)
     55 
     56         try:
---> 57             result = self.fn(*self.args, **self.kwargs)
     58         except BaseException as exc:
     59             self.future.set_exception(exc)

/datasets/jkirkham/miniconda/envs/rapids15dev/lib/python3.8/site-packages/cudf/core/abc.py in _ensure_maybe_on_device(is_cuda, buf)
     44     if is_cuda:
     45         with _get_tls_stream() as stream:
---> 46             dbuf = rmm.DeviceBuffer(size=buf.nbytes, stream=stream.ptr)
     47             cupy.asarray(dbuf).set(numpy.asarray(buf), stream=stream)
     48             return dbuf

rmm/_lib/device_buffer.pyx in rmm._lib.device_buffer.DeviceBuffer.__cinit__()

RuntimeError: CUDA error at: ../include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp365: cudaErrorInvalidResourceHandle invalid resource handle
```

</details>